src/SAWScript/MGU.hs

```diff
@@ -67,44 +66,146 @@ listSubst = Subst . M.fromList
 
 -- Most General Unifier {{{
 
-failMGU :: String -> Either String a
-failMGU = Left
-
-assert :: Bool -> String -> Either String ()
-assert b msg = unless b $ failMGU msg
-
-mgu :: LName -> Type -> Type -> Either String Subst
-mgu m (LType _ t) t2 = mgu m t t2
-mgu m t1 (LType _ t) = mgu m t1 t
-mgu m (TyUnifyVar i) t2 = bindVar m i t2
-mgu m t1 (TyUnifyVar i) = bindVar m i t1
-mgu m r1@(TyRecord ts1) r2@(TyRecord ts2) = do
-  assert (M.keys ts1 == M.keys ts2) $
-    "Record field type mismatch. Expected " ++ pShow r1 ++ " but got " ++ pShow r2 ++ " at " ++ show m
-  mgus m (M.elems ts1) (M.elems ts2)
-mgu m (TyCon tc1 ts1) (TyCon tc2 ts2)
-  | tc1 == tc2 = (mgus m ts1 ts2)
+{-
+ - Error reporting.
+ -
+ - When we find a mismatch, we have potentially recursed arbitrarily
+ - deeply into the original type. We need to print the specific types
+ - we trip on (this is important if they are e.g. elements in a large
+ - system of nested records and typles) but we also want to print the
+ - rest of the original context as well.
+ -
+ - Therefore, we start with an initial descriptive message plus (in
+ - most cases) a pair of expected and found types. Once we fail, we
+ - add more expected/found type pairs on the way out of the recursion,
+ - so we print every layer of the type.
```

Comment:
Do you think it is worth adding a cutoff for the number of layers that we track? I wonder how large the error messages will get in practice.

Reply:
I suspect if we get a deep one that we'll want to keep all the layers anyway. It's also not obvious where to cut if you wanted to cut some. I made a contrived example and got this message out:
```
test2.saw:19:5-19:35: Type mismatch.
    Mismatch of type constructors. Expected: String but got Int
    Expected: String
    Found:    Int
    Expected: (Int, String)
    Found:    (Int, Int)
    Expected: (Int, String) -> Int
    Found:    (Int, Int) -> Int
    Expected: {a : Int, b : (Int, String) -> Int}
    Found:    {a : Int, b : (Int, Int) -> Int}
    Expected: ({a : Int, b : (Int, String) -> Int}, Int)
    Found:    ({a : Int, b : (Int, Int) -> Int}, Int)
    Expected: ({a : Int, b : (Int, String) -> Int}, Int) -> ({a : Int, b : (Int, String) -> Int}, Int)
    Found:    (t.1, Int) -> ({a : Int, b : (Int, Int) -> Int}, Int)
    Expected: {c : ({a : Int, b : (Int, String) -> Int}, Int) -> ({a : Int, b : (Int, String) -> Int}, Int)}
    Found:    {c : (t.1, Int) -> ({a : Int, b : (Int, Int) -> Int}, Int)}
    within "buzz" (test2.saw:19:5-19:9)
```
Example is as follows:
```
typedef MyStruct = {
   a: Int,
   b: ((Int, String) -> Int)
};

typedef OtherStruct = {
   c: ((MyStruct, Int) -> (MyStruct, Int))
};

let foo ((x: Int), (y: Int)) = y;

let bar ((m: MyStruct), (k: Int)) =
   let x' = { a = m.a, b = foo } in
   (x', k);

let baz () = { c = bar };

let buzz () : OtherStruct = baz ();
```
which is contrived but not absolutely ridiculous.

The cases that prompted me to write this were not this deep (3-4) but involved large structs where the individual found/expected lines were lengthy.

Reply:
By which I guess I mean: if you can think of a good heuristic for cutting some of the layers I'm happy to stuff it in, but I can't, and I suspect just bounding the number is not going to be a good one. Plus, in the example above, even though there are seven layers it's not horrific. If the types involved were larger (e.g. more struct members) then the lines would be longer and there'd be more text, but the more struct members there are the more important it is to pick out the one that's mismatched. So I think that balances out.

IDK, my inclination is to leave it and if it turns out to be annoyingly verbose in practice move up the priority of stuff like teaching it to report which argument of which function it's talking about.

Reply:
I don't have a heuristic in mind. I was simply curious if you had given the matter some thought, and you've certainly thought about this more than I have, so I'm content with the current approach.

That being said, seeing the example that you posted makes me wonder if we shouldn't add a little more whitespace so that it looks like this instead:

```
test2.saw:19:5-19:35: Type mismatch.
    Mismatch of type constructors. Expected: String but got Int
    
    Expected: String
    Found:    Int
    
    Expected: (Int, String)
    Found:    (Int, Int)
    
    Expected: (Int, String) -> Int
    Found:    (Int, Int) -> Int
    
    Expected: {a : Int, b : (Int, String) -> Int}
    Found:    {a : Int, b : (Int, Int) -> Int}
    
    Expected: ({a : Int, b : (Int, String) -> Int}, Int)
    Found:    ({a : Int, b : (Int, Int) -> Int}, Int)
    
    Expected: ({a : Int, b : (Int, String) -> Int}, Int) -> ({a : Int, b : (Int, String) -> Int}, Int)
    Found:    (t.1, Int) -> ({a : Int, b : (Int, Int) -> Int}, Int)
    
    Expected: {c : ({a : Int, b : (Int, String) -> Int}, Int) -> ({a : Int, b : (Int, String) -> Int}, Int)}
    Found:    {c : (t.1, Int) -> ({a : Int, b : (Int, Int) -> Int}, Int)}
    
    within "buzz" (test2.saw:19:5-19:9)
```

Otherwise, the `Expected`/`Found` lines kind of blend into each other, which requires you to stare a bit to figure out how to group them.

Also, I wonder if we should just drop the `Expected: String but got Int` part of the error message. After the changes in this PR, the same information is now contained in this part of the error:

```
    Expected: String
    Found:    Int
```

Which makes the `Expected: String but got Int` part a bit redundant.

Reply:
The "Mismatch of type constructors" line prints specifically the type constructors, so it's not always the same (though it is here)... not sure if there are any cases where that will add clarity/be helpful, possibly not.

I think I agree on the newlines, I'll do that.